x-pack/plugins/apm/common/idx.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/**
+ * DeepRequiredArray
+ * Nested array condition handler
+ */
+interface DeepRequiredArray<T> extends Array<DeepRequired<T>> {}
+
+/**
+ * DeepRequiredObject
+ * Nested object condition handler
+ */
+type DeepRequiredObject<T> = { [P in keyof T]-?: DeepRequired<T[P]> };
+
+/**
+ * Function that has deeply required return type
+ */
+type FunctionWithRequiredReturnType<
+  T extends (...args: any[]) => any
+> = T extends (...args: infer A) => infer R
+  ? (...args: A) => DeepRequired<R>
+  : never;
+
+/**
+ * DeepRequired
+ * Required that works for deeply nested structure
+ */
+type DeepRequired<T> = NonNullable<T> extends never
```

Comment:
What does this `DeepRequired` mean for the `idx` accessor?

Reply:
It recursively make optional properties required, so typescript won't complain when accessing props that are maybe undefined.